map, flatMap, filter on Metadata

paulp opened the pull request.

paulp committed:
map, flatMap, filter on Metadata

paulp commented:
Only intended as a basis for discussion, not so much as a finished product.

paulp committed:
Simplify the extractor.

There's no real way to use the key generically, you wind up
having to match on the data value anyway.

EECOLOR commented on core/src/main/scala/api/metadata.scala:
```diff
@@ -39,9 +16,13 @@ object Example {
   private val untypedMap                         = attributes.foldLeft(Map[Key[_], Any]())(_ + _.pair)
   private def untypedAs[A]()(implicit z: Key[A]) = untypedMap(z).asInstanceOf[A]
 
+  def transformAttributes(f: Vector[Attribute] => Vector[Attribute]): Metadata = new Metadata(f(attributes))
```
I like this name

EECOLOR commented:
I have been staring at the rest. And I am convinced we should add methods as we use them. We could spend ages adding and discussing methods that end up not being used.

I think it's great if we have these methods on stand-by though.

EECOLOR commented:
Oh, and I like the `Only[A]` concept. That can prove very valuable. Not sure about the methods and their signatures until we have a few use cases

paulp committed:
Makes idfs more identity.

Timestamps, uids.

Makes follow vs. nofollow explicit and consistent with java.nio usage.

paulp commented:
I might have to stop for now but I pushed my work. In past work I've found a huge source of bugs is confusion about what will happen with symbolic links. Basically every operation where it makes a difference should be forced to either "follow" or "nofollow" to call the method. That's how I arranged it. It also improves the java.nio.file API by a lot, because there's only one "LinkOption" possible. Talk about overabstraction.

paulp committed:
Get rid of the non-instance Files forwarders.

They work a lot better the other way.

paulp committed:
reversefs.

A one-line filesystem which reverses the bytes of a file.
Novelty act, but illustrative.

% ./bin/reversefs /private/etc /mnt
...
% head -n10 /mnt/group

:993:*:hss_ssecca.elppa.moc
:893:*:gnirahsneercs_ssecca.elppa.moc
:793:*:yeknoisses_ssecca.elppa.moc
:693:*:delbasid_ssecca.elppa.moc
:593:*:ptf_ssecca.elppa.moc
:542:*:sirtsa_
:442:*:dycilopyalpsid_
:342:*:degarotslrusn_
:242:*:dnoisseslrusn_

paulp commented:
Restructuring.
